lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,546 @@
-                        return {};
-                      });
-              if (childVal && type_isa<FIRRTLType>(childVal.getType()))
-                children.push_back(childVal);
-            }
-          }
-          for (auto childVal : children) {
-            // This childVal can be ignored, if
-            // It is a Register or a subfield of a register.
-            if (!visited.contains(childVal))
-              dfsStack.push_back(childVal);
-            // If the childVal is a sub, then check if it aliases with any of
-            // the predecessors (the visiting set).
-            if (visiting.contains(childVal)) {
-              // Comb Cycle Detected !!
-              reportLoopFound(childVal, visiting);
-              return failure();
-            }
-          }
-          // child nodes added, continue the DFS
-          if (dfsSize != dfsStack.size())
-            continue;
-        }
-        // FieldRef is an SCC root node, pop the visiting stack to remove the
-        // nodes that are no longer active predecessors, that is their sub-tree
-        // is already explored. All the Values reachable from `dfsVal` have been
-        // explored, remove it and its children from the visiting stack.
-        visiting.popUntilVal(dfsVal);
-
-        auto popped = dfsStack.pop_back_val();
-        (void)popped;
-        LLVM_DEBUG({
-          llvm::dbgs() << "\n dfs popped :"
-                       << getFieldName(FieldRef(popped, 0)).first;
-          dump();
-        });
-      }
-    }
-
-    return success();
+    constructReachingDefGraph(module);
+    return dfsTraverse(reachingDefGraph);
   }
 
-  // Preprocess the module ops to get the
-  // 1. roots for DFS traversal,
-  // 2. FieldRef corresponding to each Value.
-  void preprocess(SmallVector<Value> &worklist) {
-    // All the input ports are added to the worklist.
-    for (BlockArgument arg : module.getArguments()) {
-      auto argType = type_cast<FIRRTLType>(arg.getType());
-      if (type_isa<RefType>(argType))
+  void constructReachingDefGraph(FModuleOp module) {
+    LLVM_DEBUG(llvm::dbgs() << "\n Module :" << module.getName());
+
+    // ALl the module output ports must be added as the initial nodes to the
+    // reaching def graph.
+    for (auto port : module.getArguments()) {
+      if (module.getPortDirection(port.getArgNumber()) != Direction::Out)
         continue;
-      if (module.getPortDirection(arg.getArgNumber()) == Direction::In)
-        worklist.push_back(arg);
-      if (!argType.isGround())
-        setValRefsTo(arg, FieldRef(arg, 0));
+      auto portNode = getOrAddNode(port);
+      auto portField = reachingDefGraph[portNode].first;
+      // If aggregate type, then add all the ground type fields to the graph.
+      if (auto bType = getBaseType(port.getType()))
+        if (!bType.isGround())
+          walkGroundTypes(bType, [&](uint64_t index, FIRRTLBaseType t) {
+            getOrAddNode(portField.getSubField(index));
+          });
```

Comment:
1. You create a node for fieldID 0, and then for each leaf field of the aggregate, but not intermediate aggregates in nested aggregates. E.g. `{a : {b: {c : UInt<1>}}}`, you create a node for `a` and `c` but not `b`.  Seems a little fishy?
1. `walkGroundTypes` calls `getBaseType()` internally, no need to call it here.
1. Since we are stripping out the output ports, should this be sensitive to flipped fields in aggregates?
1. Walk ground types works fine if the type is not an aggregate, so the code could maybe be simplified to something like 
```
    for (auto port : module.getArguments()) {
      if (module.getPortDirection(port.getArgNumber()) != Direction::Out)
        continue;
      walkGroundTypes(port.getType(), [&](uint64_t index, FIRRTLBaseType t) {
        getOrAddNode(FieldIndex(port, index));
      });
    }
```

Reply:
1. The intermediate fields need not be recorded explicitly, since the dataflow is always tracked with respect to leaf ground types. If the intermediate field is accessed then it will be automatically added to the graph, and then all the leaf nodes corresponding to the intermediate node will be updated with the edges on the graph.

Reply:
3. This implementation does not handle flip types, since it seemed like they cannot reach this pass in the FIRRTL pipeline. I will look into it and maybe handle it in a future PR, if this complicates the logic too much.
4. Thanks for the code snippet, used it.